src/rp2_common/pico_rand/rand.c

```diff
@@ -0,0 +1,303 @@
+   small-state generators. It is extremely (sub-ns) fast and it passes all
+   tests we are aware of, but its state space is large enough only for
+   mild parallelism.
+
+   For generating just floating-point numbers, xoroshiro128+ is even
+   faster (but it has a very mild bias, see notes in the comments).
+
+   The state must be seeded so that it is not everywhere zero. If you have
+   a 64-bit seed, we suggest to seed a splitmix64 generator and use its
+   output to fill s.
+*/
+static inline uint64_t rotl(const uint64_t x, int k) {
+    return (x << k) | (x >> (64 - k));
+}
+
+static uint64_t xoroshiro128ss(void) {
+
+    const uint64_t s0 = rand_storage.r[0];
+    uint64_t s1 = rand_storage.r[1];
+
+    // Because the state is *modified* outside of this function, there is a
+    // 1 in 2^128 chance that it could be all zeroes (which is not allowed).
+    if (s0 == 0 && s1 == 0) {
+        s1 = time_us_64();   // infeasibly 0
+    }
+
+    const uint64_t result = rotl(s0 * 5, 7) * 9;
+
+    s1 ^= s0;
+    rand_storage.r[0] = rotl(s0, 24) ^ s1 ^ (s1 << 16); // a, b
+    rand_storage.r[1] = rotl(s1, 37); // c
+
+    return result;
+}
+
+
+// Note: This can take many tens of milliseconds to execute
+static uint64_t sdbm_hash64_sram_under_lock(uint64_t hash) {
+
+    for (unsigned int i = RAM_HASH_START; i < RAM_HASH_END; i++) {
+
+        uint8_t c = *(uint8_t *) i;
+        hash = (uint64_t) c + (hash << 6) + (hash << 16) - hash;
+    }
+    return hash;
+}
+
+
+static uint32_t frequency_count_rosc_under_lock(void) {
+
+    fc_hw_t *fc = &clocks_hw->fc0;
+
+    // If frequency counter is running need to wait for it. It runs even if the source is NULL
+    while(fc->status & CLOCKS_FC0_STATUS_RUNNING_BITS) {
+        tight_loop_contents();
+    }
+
+    // Set reference freq
+    fc->ref_khz = clock_get_hz(clk_ref) / 100;
+
+    fc->interval = 100;
+
+    // No min or max
+    fc->min_khz = 0;
+    fc->max_khz = 0xfffffffful;
+
+    // Set SRC which automatically starts the measurement
+    fc->src = CLOCKS_FC0_SRC_VALUE_ROSC_CLKSRC;
+
+    while(!(fc->status & CLOCKS_FC0_STATUS_DONE_BITS)) {
+        tight_loop_contents();
+    }
+
+    // Return the result.  Note: we don't shift right
+    return fc->result;
+}
+
+
+// check that the ROSC is running but that the processors are NOT running from it
+static inline void check_rosc_asserts(void) {
+    hard_assert(rosc_hw->status & ROSC_STATUS_ENABLED_BITS);
+    hard_assert((clocks_hw->clk[clk_sys].ctrl & CLOCKS_CLK_SYS_CTRL_AUXSRC_BITS) !=
+           (CLOCKS_CLK_SYS_CTRL_AUXSRC_VALUE_ROSC_CLKSRC << CLOCKS_CLK_SYS_CTRL_AUXSRC_LSB));
+}
+
+
+static uint32_t measure_rosc_freq_under_lock(void) {
+
+    uint32_t rosc_freq;
+
+    uint32_t snapshot = rosc_hw->ctrl;
+
+    check_rosc_asserts();
+
+    // speed up rosc, wait until stable
+    rosc_hw->ctrl = (ROSC_CTRL_ENABLE_VALUE_ENABLE   << ROSC_CTRL_ENABLE_LSB) |
+                    (ROSC_CTRL_FREQ_RANGE_VALUE_HIGH << ROSC_CTRL_FREQ_RANGE_LSB);
```

Comment:
what does speeding up here do for the random number generator?
does the sped-up frequency only go into initialisation as a tiny bit of startup entropy?
could we make this optional? there might be stuff running from the rosc (other than clk_sys).

Reply:
yeah, i removed this